sympy/physics/optics/medium.py

```diff
@@ -69,23 +70,31 @@ class Medium(Symbol):
     """
 
     def __new__(cls, name, permittivity=None, permeability=None, n=None):
-        obj = super().__new__(cls, name)
-        obj._permittivity = sympify(permittivity)
-        obj._permeability = sympify(permeability)
-        obj._n = sympify(n)
+        if not isinstance(name, Str):
+            name = Str(name)
+
         if n is not None:
             if permittivity is not None and permeability is None:
-                obj._permeability = n**2/(c**2*obj._permittivity)
+                permeability = n**2/(c**2*permittivity)
             if permeability is not None and permittivity is None:
-                obj._permittivity = n**2/(c**2*obj._permeability)
+                permittivity = n**2/(c**2*permeability)
             if permittivity is not None and permittivity is not None:
-                if abs(n - c*sqrt(obj._permittivity*obj._permeability)) > 1e-6:
-                   raise ValueError("Values are not consistent.")
+                expr = abs(n - c*sqrt(permittivity*permeability))
+                expr = expr.subs({meter: 1, second: 1})
+                if len(expr.free_symbols) == 0 and expr > 1e-6:
+                    raise ValueError("Values are not consistent.")
```

Comment:
What if `n is not None`, `permeability is None` and `permittivity is None`? I don't think it is possible to compute the values for `permeability` and `permittivity` in that case (you have two unknowns but one equation), but this is not handled explicitly here. I think it will end up with a `None` in both `permeability` and `permittivity` when it passes them to `_sympify` which will cause an error.

Reply:
Line 77-85 should also be an if-elif block since the options are mutually exclusive.

Reply:
Regarding your first suggestion, in the case you are talking about, how do you suggest this should be handled? Should it immediately raise an rrror? 
Also, the version of lines 77-85 in the original code does not have `if-elif` blocks. I can still change that though.

Reply:
Yes, it should raise an error to tell the user it is using a combination of inputs that can't be used, rather than fail somewhere where the error might not be easy to understand. 

I think the `elif` would be more efficient.

Reply:
Ok. Are there any other changes you suggest I should make in this section of the code? I will make all of them in a single commit to avoid too many commits with small changes in each commit.

Reply:
If `n is None` and only one of `permittivity` and `permeability` `is None` is also not handled.

Reply:
Should that also produce an error since it's impossible to calculate unique values for both unknowns? Or should it use the default value for whichever of `permittivity` and `permeability` is `None` and then calculate `n` using that value?

Reply:
Other than that it looks good.

Reply:
I think an error is better, since it currently would also give an error at some point, just not at a place where a user can easily understand what was done wrong.

Reply:
Does it look ok now? If so, I will resolve this conversation